File: backend/Entities/TicketNumberTemplate.cs
using System;
using System.ComponentModel.DataAnnotations;

namespace SapBasisPulse.Api.Entities
{
    public class TicketNumberTemplate
    {
        public Guid Id { get; set; }
        
        [Required]
        [MaxLength(100)]
        public string Name { get; set; } = string.Empty;
        
        [MaxLength(500)]
        public string? Description { get; set; }
        
        // Template pattern using placeholders like {SupportType}, {Category}, {SubType}, {Date}, {Sequence}
        [Required]
        [MaxLength(200)]
        public string Template { get; set; } = string.Empty;
        
        // Optional: Specific support type this template applies to
        public Guid? SupportTypeId { get; set; }
        public SupportType? SupportType { get; set; }
        
        // Optional: Specific category this template applies to  
        public Guid? SupportCategoryId { get; set; }
        public SupportCategory? SupportCategory { get; set; }
        
        // Optional: Specific sub-option this template applies to
        public Guid? SupportSubOptionId { get; set; }
        public SupportSubOption? SupportSubOption { get; set; }
        
        // Priority for template selection (higher number = higher priority)
        public int Priority { get; set; } = 0;
        
        // Whether this template is active
        public bool IsActive { get; set; } = true;
        
        // Whether this is the default template (fallback)
        public bool IsDefault { get; set; } = false;
        
        // Sequence counter for generating unique numbers
        public int CurrentSequence { get; set; } = 0;
        
        // Date format for {Date} placeholder (e.g., "yyyy-MM-dd", "yyyyMMdd")
        [MaxLength(50)]
        public string DateFormat { get; set; } = "yyyy-MM-dd";
        
        // Sequence format for {Sequence} placeholder (e.g., "0000" for 4-digit padding)
        [MaxLength(20)]
        public string SequenceFormat { get; set; } = "0000";
        
        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
        public DateTime? UpdatedAt { get; set; }
        
        // Audit fields
        public Guid CreatedByUserId { get; set; }
        public User CreatedByUser { get; set; } = null!;
        
        public Guid? UpdatedByUserId { get; set; }
        public User? UpdatedByUser { get; set; }
    }
}

File: frontend/src/types/ticketNumberTemplate.ts
// API types for ticket number templates
export interface TicketNumberTemplate {
  id: string;
  name: string;
  description?: string;
  template: string;
  supportTypeId?: string;
  supportTypeName?: string;
  supportCategoryId?: string;
  supportCategoryName?: string;
  supportSubOptionId?: string;
  supportSubOptionName?: string;
  priority: number;
  isActive: boolean;
  isDefault: boolean;
  currentSequence: number;
  dateFormat: string;
  sequenceFormat: string;
  createdAt: string;
  updatedAt?: string;
  createdByUserName: string;
  updatedByUserName?: string;
}

export interface CreateTicketNumberTemplate {
  name: string;
  description?: string;
  template: string;
  supportTypeId?: string;
  supportCategoryId?: string;
  supportSubOptionId?: string;
  priority: number;
  isActive: boolean;
  isDefault: boolean;
  dateFormat: string;
  sequenceFormat: string;
}

export interface UpdateTicketNumberTemplate {
  name: string;
  description?: string;
  template: string;
  supportTypeId?: string;
  supportCategoryId?: string;
  supportSubOptionId?: string;
  priority: number;
  isActive: boolean;
  isDefault: boolean;
  dateFormat: string;
  sequenceFormat: string;
}

export interface TicketNumberTemplatePreview {
  template: string;
  dateFormat: string;
  sequenceFormat: string;
  supportTypeId?: string;
  supportCategoryId?: string;
  supportSubOptionId?: string;
}

export interface TicketNumberPreviewResult {
  previewNumber: string;
  isValid: boolean;
  errorMessage?: string;
}

export interface TemplateFormData {
  name: string;
  description: string;
  template: string;
  supportTypeId: string;
  supportCategoryId: string;
  supportSubOptionId: string;
  priority: number;
  isActive: boolean;
  isDefault: boolean;
  dateFormat: string;
  sequenceFormat: string;
}

// Available template placeholders
export const TEMPLATE_PLACEHOLDERS = [
  { value: '{SupportType}', label: 'Support Type', description: 'Short name of the support type (e.g., SAPRISE, SAPGROW)' },
  { value: '{Category}', label: 'Category', description: 'Short name of the category (e.g., BASIS, DB, OS)' },
  { value: '{SubType}', label: 'Sub Type', description: 'Short name of the sub-option (e.g., SR, INC)' },
  { value: '{Date}', label: 'Date', description: 'Current date in specified format (e.g., 2025-10-05)' },
  { value: '{Sequence}', label: 'Sequence', description: 'Auto-incrementing number in specified format (e.g., 0001)' }
] as const;

// Common date format options
export const DATE_FORMAT_OPTIONS = [
  { value: 'yyyy-MM-dd', label: '2025-10-05', description: 'Full date with dashes' },
  { value: 'yyyyMMdd', label: '20251005', description: 'Full date without separators' },
  { value: 'yyyy-MM', label: '2025-10', description: 'Year and month only' },
  { value: 'yyyy', label: '2025', description: 'Year only' },
  { value: 'MMdd', label: '1005', description: 'Month and day only' }
] as const;

// Common sequence format options
export const SEQUENCE_FORMAT_OPTIONS = [
  { value: '0', label: '1', description: 'Simple number (1, 2, 3...)' },
  { value: '00', label: '01', description: '2-digit padded (01, 02, 03...)' },
  { value: '000', label: '001', description: '3-digit padded (001, 002, 003...)' },
  { value: '0000', label: '0001', description: '4-digit padded (0001, 0002, 0003...)' },
  { value: '00000', label: '00001', description: '5-digit padded (00001, 00002, 00003...)' }
] as const;